content/posts/linux/Desktop-Linux-Hardening.md

```diff
@@ -249,19 +249,44 @@ If you are using KickSecure or Whonix, most of these hardening have already been
+In this section we succinctly present the parameters used by Kicksecure as those are more regularly updated though strongly recommend reading through Madaidan's guide.
 
+- CPU mitigations
 ```
-slab_nomerge init_on_alloc=1 init_on_free=1 page_alloc.shuffle=1 pti=on vsyscall=none debugfs=off oops=panic module.sig_enforce=1 lockdown=confidentiality mce=0 quiet loglevel=0 spectre_v2=on spec_store_bypass_disable=on tsx=off tsx_async_abort=full,nosmt mds=full,nosmt l1tf=full,force nosmt=force kvm.nx_huge_pages=force randomize_kstack_offset=on
+spectre_v2=on spec_store_bypass_disable=on l1tf=full,force mds=full,nosmt tsx=off tsx_async_abort=full, mds=full,nosmt kvm.nx_huge_pages=force nosmt=force l1d_flush=on mmio_stale_data=full,nosmt
 ```
 
-Note that [SMT](https://en.wikipedia.org/wiki/Simultaneous_multithreading) is disabled due to it being the cause of various security vulnerabilities. Also, on rpm-ostree based distributions, you should set the kernel parameters using `rpm-ostree kargs` rather than messing with `GRUB` configurations directly.
+[SMT](https://en.wikipedia.org/wiki/Simultaneous_multithreading) is disabled due to it being the cause of various security vulnerabilities. Also, on rpm-ostree based distributions, you should set the kernel parameters using `rpm-ostree kargs` rather than messing with `GRUB` configurations directly. As an aside, one should keep in mind that despite the clear security benefits of disabling SMT, the very popular `linux-hardened` kernel for Arch Linux does not disable it by default given the [large potential performance costs](https://github.com/anthraxx/linux-hardened/issues/37#issuecomment-619597365). You should determine your own desired level of risk mitigation and if you choose to keep SMT enabled, simply remove all occurrences of `nosmt` and `nosmt=force` from the above parameters.
+
+- Kernel
+```
+slab_nomerge init_on_alloc=1 init_on_free=1 pti=on vsyscall=none page_alloc.shuffle=1 randomize_kstack_offset=on extra_latent_entropy debugfs=off oops=panic quiet loglevel=0
+```
+
+Kicksecure does not enforce either `module.sig_enforce=1` or ` lockdown=confidentiality` by default as they lead a lot of hardware compatibility issues, consider enabling these if possible on your system. Additionally, `mce=0` is also no [no longer](https://forums.whonix.org/t/kernel-hardening/7296/493) used.
+
+- Entropy generation
+```
+random.trust_cpu=off random.trust_bootloader=off
+```
+
+As sources of initial entropy at boot, both the CPU and bootloader should be [distrusted](https://lkml.org/lkml/2022/6/5/271). For CPUs, the RBRAND instructions set is [impossible to audit](https://madaidans-insecurities.github.io/guides/linux-hardening.html#rdrand), and moving forward as a precaution, the bootloader should be treated identically. Note that both of these kernel parameters will increase boot time.
```

Comment:
I think there is a misunderstanding regarding what `distrust` actually does. It does not discard the entropy generated by these sources, in fact it pools it as per usual and then utilises it in an indistinguishable manner. 

With `distrust`, the system is forced to obtain more entropy at boot due simply because it does not not credit the CPU/bootloader entropy into the total amount required. I have yet to come across any evidence that proves having less entropy is superior to obtaining more for hardening purposes.

The singular downside to this approach is an increased boot time as more entropy must be sourced. This can be particularly slow on VMs.

Regarding `rng_core.default_quality=1000`, the only effect the will have is decreasing boot times as the entropy generated by these sources will be credited to the required amount. But credited does not mean included, [it was always included](https://systemd.io/RANDOM_SEEDS/).

Finally, when it comes to risk mitigation, systems with “known flawed RNG implementation” are by definition known. What of systems that have unknown flaws? At the cost of lengthier boot you can be better shielded against faulty RNG implementation.

See also the discussion when [I introduced distrusting the bootloader](https://forums.whonix.org/t/kernel-hardening/7296/465) in Kicksecure/Whonix ~5 months ago.

Reply:
In your original text, you wrote "both the CPU and bootloader should be distrusted" implying the plain English definition of the term. If you mean something different (which in hindsight I can see), you should make it clear — perhaps italicize the term and clarify the definition in parentheses or elsewhere in context.

Regarding the actual recommendations. Perhaps I am misunderstanding, so please explain what exactly happens with `random.trust_cpu=off random.trust_bootloader=off`.
- Is entropy sourced from other sources where it otherwise wouldn't be?
- Are other sources providing significantly more entropy than they would normally be?
  - If so, what makes those sources any more trustworthy than the CPU and EFI if they now have a greater contribution to the entropy pool?
- What specifically are those other sources, if any? The set of all hardware RNGs on the system?

Basically, unless these kernel parameters are actually _adding more sources to the mix_ rather than merely changing the weights of existing sources, I am still not convinced these settings are useful (though I am much more receptive to discrediting the EFI knowing how badly many of them are programmed).

Reply:
Maybe we can just word it as "Set distrust to increase entropy"?

Reply:
````suggestion
```
random.trust_cpu=off random.trust_bootloader=off
```

Some implementations of the RDRAND instruction (by which the CPU offers a random number generator to the OS) have proven to be [vulnerable](https://en.wikipedia.org/wiki/RDRAND#Security_issues) or [outright defective](https://arstechnica.com/gadgets/2019/10/how-a-months-old-amd-microcode-bug-destroyed-my-weekend/). RDRAND is also impossible to audit, being part of the CPU itself.

As a precaution for the integrity of cryptographic operations, the CPU and bootloader should not be used as _credited_ entropy sources. Note that this change will increase boot time.

Further reading:

- [systemd: Random Seeds](https://systemd.io/RANDOM_SEEDS/)
- [Madaidan: RDRAND](https://madaidans-insecurities.github.io/guides/linux-hardening.html#rdrand)
- [Linux kernel mailing list](https://lore.kernel.org/lkml/20220605171539.417872-1-Jason@zx2c4.com/T/)
- [Hacker News discussion](https://news.ycombinator.com/item?id=33223232)
- [NixOS discussion](https://github.com/NixOS/nixpkgs/pull/165355) (also cites many additional sources)
````

Reply:
For future reference, this discussion was partially split to https://github.com/PrivSec-dev/privsec.dev/pull/81#issuecomment-1367535519 and https://github.com/PrivSec-dev/privsec.dev/pull/81#issuecomment-1374423909.

Reply:
````suggestion
```
random.trust_cpu=off random.trust_bootloader=off
```

Some implementations of the RDRAND instruction (by which the CPU offers a random number generator to the OS) have proven to be [vulnerable](https://en.wikipedia.org/wiki/RDRAND#Security_issues) or [outright defective](https://arstechnica.com/gadgets/2019/10/how-a-months-old-amd-microcode-bug-destroyed-my-weekend/). RDRAND is also impossible to audit, being part of the CPU itself.

As a precaution for the integrity of cryptographic operations, the CPU and bootloader should not be used as _credited_ entropy sources. Note that this change will increase boot time.

Further reading:

- [systemd: Random Seeds](https://systemd.io/RANDOM_SEEDS/)
- [Madaidan: RDRAND](https://madaidans-insecurities.github.io/guides/linux-hardening.html#rdrand)
- [Linux kernel mailing list](https://lore.kernel.org/lkml/20220605171539.417872-1-Jason@zx2c4.com/T/)
- [Hacker News discussion](https://news.ycombinator.com/item?id=33223232)
- [NixOS discussion](https://github.com/NixOS/nixpkgs/pull/165355) (also cites many additional sources)
````

@raja-grewal I added additional background information here. I believe extensive background information is particularly important here to avoid reactions like my own initial reaction. Please let me know your thoughts and commit/edit these changes as you see fit.